library/ssl_ticket.c

```diff
@@ -254,6 +249,18 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
     ctx->keys[1].key_type = key_type;
     ctx->keys[1].key_bits = key_bits;
 #else
+    cipher_info = mbedtls_cipher_info_from_type( cipher );
+
+    if( mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_GCM &&
+        mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_CCM &&
+        mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_CHACHAPOLY )
+    {
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    if( mbedtls_cipher_info_get_key_bitlen( cipher_info ) > 8 * MAX_KEY_BYTES )
```

Comment:
Not a blocker: the less code is different between `USE_PSA_CRYPTO` and `!USE_PSA_CRYPTO` the better. In particular, the key length check can be moved out: declare `key_bits` unconditionally, set it differently depending on USE_PSA, and check it outside the `#if`.

Reply:
I see but with PSA, key_bits is set by `mbedtls_ssl_cipher_to_psa()`, and with MBedTLS it can only be retrieved after `mbedtls_cipher_info_from_type()`.

This would mean I would need to move the `mbedtls_ssl_cipher_to_psa()/mbedtls_cipher_info_from_type()` at the beginning, then add a size check in between ?

```c
int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
    mbedtls_cipher_type_t cipher,
    uint32_t lifetime )
{
    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
    size_t key_bits;

#if defined(MBEDTLS_USE_PSA_CRYPTO)
    psa_algorithm_t alg;
    psa_key_type_t key_type;
#else
    const mbedtls_cipher_info_t *cipher_info;
#endif /* MBEDTLS_USE_PSA_CRYPTO */

#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( mbedtls_ssl_cipher_to_psa( cipher, TICKET_AUTH_TAG_BYTES,
                                   &alg, &key_type, &key_bits ) != PSA_SUCCESS )
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );

    if( PSA_ALG_IS_AEAD( alg ) == 0 )
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
#else
    cipher_info = mbedtls_cipher_info_from_type( cipher );

    if( mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_GCM &&
        mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_CCM &&
        mbedtls_cipher_info_get_mode( cipher_info ) != MBEDTLS_MODE_CHACHAPOLY )
    {
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
    }

    key_bits = mbedtls_cipher_info_get_key_bitlen( cipher_info );
#endif /* MBEDTLS_USE_PSA_CRYPTO */

    if( key_bits > 8 * MAX_KEY_BYTES )
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );

    ctx->f_rng = f_rng;
    ctx->p_rng = p_rng;

    ctx->ticket_lifetime = lifetime;

#if defined(MBEDTLS_USE_PSA_CRYPTO)
    ctx->keys[0].alg = alg;
    ctx->keys[0].key_type = key_type;
    ctx->keys[0].key_bits = key_bits;

    ctx->keys[1].alg = alg;
    ctx->keys[1].key_type = key_type;
    ctx->keys[1].key_bits = key_bits;
#else
    if( ( ret = mbedtls_cipher_setup( &ctx->keys[0].ctx, cipher_info ) ) != 0 )
        return( ret );

    if( ( ret = mbedtls_cipher_setup( &ctx->keys[1].ctx, cipher_info ) ) != 0 )
        return( ret );
#endif /* MBEDTLS_USE_PSA_CRYPTO */

    if( ( ret = ssl_ticket_gen_key( ctx, 0 ) ) != 0 ||
        ( ret = ssl_ticket_gen_key( ctx, 1 ) ) != 0 )
    {
        return( ret );
    }

    return( 0 );
}
```

The problem is we can't use `PSA_BYTES_TO_BITS()` anymore, except adding a const `key_bits_max set` differently for PSA, but it makes thing really complicated.

Reply:
Looking just at the code here and not in comparison with the previous code, yes, that's what I had in mind: set `key_bits` differently depending on USE_PSA, but have the check in code that's always executed.

Technically you can actually use `PSA_BYTES_TO_BITS` if you include `psa/crypto.h`, regardless of whether `MBEDTLS_PSA_CRYPTO_C` is enabled. But we might want to not call any PSA macro if `MBEDTLS_PSA_CRYPTO_C` is disabled, as a style thing and to avoid accidentally relying on `MBEDTLS_PSA_CRYPTO_C` being enabled. But anyway `PSA_BYTES_TO_BITS` isn't a crucial macro, you can do the ×8 manually. (It exists mostly as a counterpart to the slightly more complex `PSA_BITS_TO_BYTES`.)